File: leetcode2025/2025_09/leetcode_1733_minimumTeachings.py
from typing import List

class Solution:
    def minimumTeachings(self, n: int, languages: List[List[int]], friendships: List[List[int]]) -> int:
        # 对于每一个朋友对(a,b)，只要两者所掌握的语言中没有交集，我们就将它标记为待处理朋友对，之后，对于所有的待处理朋友对，我们都选择同一种语言进行教学，因此，朋友对中所有的不会这一种语言的人的总数量就是一个答案。但是我们可以教他们学任意一种语言，因此循环所有的语言，都进行一次计算，取最小的答案即可
        lang = list(map(set, languages))
        lang.insert(0, {})
        teach = []
        for a,b in friendships:
            if len(lang[a].intersection(lang[b])) == 0:
            # if lang[a].isdisjoint(lang[b]):
                teach.append([a,b])

        res = float("inf")
        for k in range(1, n+1):
            ans = 0
            tmp = set()
            for a,b in teach:
                if k not in lang[a]:
                    tmp.add(a)
                if k not in lang[b]:
                    tmp.add(b)
            res = min(res, len(tmp))
        return res

if __name__ == "__main__":
    s = Solution()
    n = 2
    languages = [[1],[2],[1,2]]
    friendships = [[1,2],[1,3],[2,3]]
    print(s.minimumTeachings(n, languages, friendships))

File: leetcode2025/2025_09/leetcode_1733_minimumTeachings.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <climits>
using namespace std;

class Solution {
public:
    int minimumTeachings(int n, vector<vector<int>>& languages, vector<vector<int>>& friendships) {
        vector<unordered_set<int>> lang(languages.size());
        for (int i=0; i<languages.size(); i++){
            lang[i] = unordered_set<int>(languages[i].begin(), languages[i].end());
        }

        vector<vector<int>> teach;
        for (auto& f : friendships){
            bool flag = true;
            int a = f[0]-1, b = f[1]-1;
            for (int l : languages[a]){
                if (lang[b].count(l)){
                    flag = false;
                    break;
                }
            }
            if (flag){
                teach.push_back({a, b});
            }
        }

        int res = INT_MAX;
        for (int k=1; k<=n; k++){
            unordered_set<int> tmp;
            for (auto& t : teach){
                int a = t[0], b = t[1];
                if (lang[a].count(k)==0){
                    tmp.insert(a);
                }
                if (lang[b].count(k)==0){
                    tmp.insert(b);
                }
            }
            res = min(res, (int)tmp.size());
        }
        return res;
    }
};

int main(){
    Solution s;
    int n = 2;
    vector<vector<int>> languages = {{1},{2},{1,2}};
    vector<vector<int>> friendships = {{1,2},{1,3},{2,3}};
    cout << s.minimumTeachings(n, languages, friendships) << endl;
}
